File: java/src/com/ctre/phoenix/Mechanical/Linkage.java
package com.ctre.phoenix.Mechanical;

import com.ctre.phoenix.MotorControl.*;

public class Linkage implements com.ctre.phoenix.Signals.IOutputSignal, com.ctre.phoenix.Signals.IInvertable
{
	IMotorController _motor = null;
	IFollower[] _follower = {null, null, null};
	int _followerCount = 0;
	boolean _isInverted = false;
	
	public Linkage(IMotorController mc1)
	{
		_motor = mc1;
		Setup();
	}
	public Linkage(IMotorController mc1, IMotorController mc2)
	{
		_motor = mc1;
		_follower[_followerCount++] = mc2;
		Setup();
	}
	public Linkage(IMotorController mc1, IMotorController mc2, IMotorController mc3)
	{
		_motor = mc1;
		_follower[_followerCount++] = mc2;
		_follower[_followerCount++] = mc3;
		Setup();
	}
	public Linkage(IMotorController mc1, IMotorController mc2, IMotorController mc3, IMotorController mc4)
	{
		_motor = mc1;
		_follower[_followerCount++] = mc2;
		_follower[_followerCount++] = mc3;
		_follower[_followerCount++] = mc4;
		Setup();
	}
	
	private void Setup()
	{
		for(int i = 0; i < _followerCount; i++)
		{
			_follower[i].follow(_motor);
		}
	}
	
	public void set(double value)
	{
		if(_isInverted)
			value = -value;
		_motor.set(value);
		for(int i = 0; i < _followerCount; i++)
		{
			_follower[i].valueUpdated();
		}
	}
	
	public void SetControlMode(ControlMode.BasicControlMode mode)
	{
		_motor.setControlMode(mode);
	}
	
	public void setInverted(boolean inverted)
	{
		_isInverted = inverted;
	}
	
	public boolean getInverted()
	{
		return _isInverted;
	}
	
	public void SetVoltageRampRate(float rampRate)
    {
        _motor.setVoltageRampRate(rampRate);
    }

    public void SetVoltageCompensationRampRate(float rampRate)
    {
        _motor.setVoltageRampRate(rampRate);
    }

    public void ConfigNominalOutputVoltage(float forwardVoltage, float reverseVoltage)
    {
        _motor.configNominalOutputVoltage(forwardVoltage, reverseVoltage);
    }

    public void ConfigPeakOutputVoltage(float forwardVoltage, float reverseVoltage)
    {
        _motor.configPeakOutputVoltage(forwardVoltage, reverseVoltage);
    }

    public IMotorController getMaster()
    {
        return _motor;
    }
}

File: cpp/src/Mechanical/Linkage.cpp
//#include "ctre/phoenix/Mechanical/Linkage.h"
//
//
//namespace ctre{ namespace mechanical{
//
//Linkage::Linkage(ctre::motorcontrol::IMotorController *mc1) {
//	_motor = mc1;
//	Setup();
//}
//Linkage::Linkage(ctre::motorcontrol::IMotorController *mc1, ctre::motorcontrol::IFollower *mc2){
//	_motor = mc1;
//	_follower[_followerCount++] = mc2;
//	Setup();
//}
//Linkage::Linkage(ctre::motorcontrol::IMotorController *mc1, ctre::motorcontrol::IFollower *mc2, ctre::motorcontrol::IFollower *mc3){
//	_motor = mc1;
//	_follower[_followerCount++] = mc2;
//	_follower[_followerCount++] = mc3;
//	Setup();
//}
//Linkage::Linkage(ctre::motorcontrol::IMotorController *mc1, ctre::motorcontrol::IFollower *mc2, ctre::motorcontrol::IFollower *mc3, ctre::motorcontrol::IFollower *mc4){
//	_motor = mc1;
//	_follower[_followerCount++] = mc2;
//	_follower[_followerCount++] = mc3;
//	_follower[_followerCount++] = mc4;
//	Setup();
//}
//void Linkage::Set(double output){
//	if(_isInverted)
//		output = -output;
//	_motor->Set(output);
//}
//void Linkage::SetControlMode(ctre::motorcontrol::ControlMode::BasicControlMode mode){
//	_motor->SetControlMode(mode);
//}
//void Linkage::SetInverted(bool invert){
//	_isInverted = invert;
//}
//bool Linkage::GetInverted() const{
//	return _isInverted;
//}
//void Linkage::SetVoltageRampRate(float rampRate){
//	_motor->SetVoltageRampRate(rampRate);
//}
//void Linkage::SetVoltageCompensationRate(float rampRate){
//	_motor->SetVoltageRampRate(rampRate);
//}
//void Linkage::ConfigNominalOutputVoltage(float forwardVoltage, float reverseVoltage){
//	_motor->ConfigNominalOutputVoltage(forwardVoltage, reverseVoltage);
//}
//void Linkage::ConfigPeakOutputVoltage(float forwardVoltage, float reverseVoltage){
//	_motor->ConfigPeakOutputVoltage(forwardVoltage, reverseVoltage);
//}
//ctre::motorcontrol::IMotorController *Linkage::GetMaster(){
//	return _motor;
//}
//void Linkage::Setup(){
//	for(int i = 0; i < _followerCount; i++){
//		_follower[i]->Follow(_motor);
//	}
//}
//
//}}
